juju/charm.py

```diff
@@ -82,19 +117,145 @@ def __init__(self, framework, key, metadata):
+    Supported fields are:
+
+        * ``name``
+        * ``summary``
+        * ``description``
+        * ``maintainers``
+        * ``tags``
+        * ``terms``
+        * ``series``
+        * ``subordinate``
+        * ``min_juju_version``
+        * ``requires``
+        * ``provides``
+        * ``peers``
+        * ``relations``
+        * ``storage``
+        * ``resources``
+        * ``payloads``
+        * ``extra_bindings``
+
+    The ``maintainers``, ``tags``, ``terms``, ``series``, and ``extra_bindings``
+    attributes are all tuples of strings.  The ``requires``, ``provides``,
+    `peers``, ``relations``, ``storage``, ``resources``, and ``payloads``
+    attributes are all mappings of names to instances of the respective
+    :class:`RelationDefinition`, :class:`StorageDefinition`, :class:`ResourceDefinition`,
+    or :class:`PayloadDefinition`.
+
+    The ``relations`` attribute is a convenience accessor which includes all of
+    the ``requires``, ``provides``, and ``peers`` :class:`RelationDefinition`_
+    items.  If needed, the role of the relation definition can be obtained from
+    its ``role`` attribute.
+    """
+    def __init__(self, raw=None):
+        raw = raw or {}
+        self.name = raw.get('name', '')
+        self.summary = raw.get('summary', '')
+        self.description = raw.get('description', '')
+        maintainers = []
+        if 'maintainer' in raw:
+            maintainers.append(raw['maintainer'])
+        if 'maintainers' in raw:
+            maintainers.extend(raw['maintainers'])
+        self.maintainers = tuple(maintainers)
+        self.tags = tuple(raw.get('tags', []))
+        self.terms = tuple(raw.get('terms', []))
+        self.series = tuple(raw.get('series', []))
+        self.subordinate = raw.get('subordinate', False)
+        self.min_juju_version = raw.get('min-juju-version')
+        self.requires = {name: RelationDefinition('requires', name, rel)
+                         for name, rel in raw.get('requires', {}).items()}
+        self.provides = {name: RelationDefinition('provides', name, rel)
+                         for name, rel in raw.get('provides', {}).items()}
+        self.peers = {name: RelationDefinition('peers', name, rel)
+                      for name, rel in raw.get('peers', {}).items()}
+        self.relations = {}
+        self.relations.update(self.requires)
+        self.relations.update(self.provides)
+        self.relations.update(self.peers)
+        self.storage = {name: StorageDefinition(name, store)
+                        for name, store in raw.get('storage', {}).items()}
+        self.resources = {name: ResourceDefinition(name, res)
+                          for name, res in raw.get('resources', {}).items()}
+        self.payloads = {name: PayloadDefinition(name, payload)
+                         for name, payload in raw.get('payloads', {}).items()}
+        self.extra_bindings = tuple(raw.get('extra-bindings', {}))
+
+
+class RelationDefinition:
```

Comment:
Instead of CharmMetadata and RelationDefinition, how about we standardize on the _Meta_ suffix: _CharmMeta_, _RelationMeta_, etc.